Cargo.toml

```diff
@@ -203,7 +203,7 @@ rmp-serde = { version = "1.1.2", default-features = false, optional = true }
 rmpv = { version = "1.0.1", default-features = false, features = ["with-serde"], optional = true }
 
 # Prost / Protocol Buffers
-prost = { version = "0.12", default-features = false, features = ["std"] }
+prost = { version = "0.12", default-features = false, features = ["std", "no-recursion-limit"] }
```

Comment:
Generally, I am not a fan of unbounded defaults. See discussions:
* https://github.com/rustsec/advisory-db/blob/main/crates/prost/RUSTSEC-2020-0002.md
* https://github.com/tokio-rs/prost/issues/267
* https://github.com/rust-lang/rust/issues/79935

A better alternative here would be to reach out to prost maintainers and ask them to expose [RECURSION_LIMIT](https://github.com/tokio-rs/prost/blob/master/src/lib.rs#L31C17-L31C17) (100 seems a bit low) as configuration option. And then use re-expose it in the codec config.

Reply:
Agreed, infinite recursion could lead to runaway recursion bugs. I'd also feel better about exposing a configurable option.

Reply:
I've been looking over the code paths used, to map out what all is in play. It is not enough to get prost to expose RECURSION_LIMIT, or more likely to create a new constructor function for `DecodeContext` that takes a `recurse_count` field parameter. The `protobuf` codec uses `DynamicMessage` to do decoding, which then also would have to be extended to take a client-provided `DecodeContext` - the general idea being that Vector would create a `DecodeContext` with the desired `recurse_count`.

However, this only handles sources/sinks using the `protobuf` codec directly. The bigger issue is that the Vector-native source/sink do not (directly) use prost at all. They operate through tonic, using compile-time stubs generated from the proto-file. All byte-stream decoding calls simply call those autogenerated stubs to decode. There are no prost calls or anywhere to potentially modify or extend to use a client-provided `DecodeContext`.

I can't tell what/where/how the disk buffering encoding/decoding ultimately ties into prost, to figure out that angle.

Further, a configurable recursion limit knob in the protobuf codec and in the Vector source configs would actually have two different meanings, because the proto-generated Vector-native `Event` (or `EventMessage`) stubs currently have a 3-layer per decoded field recursion (Event -> Kind -> Value I believe?). This is effectively why the prost limit of 100 layers equates to a client-visible JSON recursion limit of 32/33 layers. Either documentation, or a multiplier, would have to be performed and remain in lockstep with the proto-definition of the Vector-native `Event` message type.

I would also tend to think runaway recursion would have to first get past the max message size limits?

Reply:
It looks like there is some movement on making the limit configurable (at build time): https://github.com/tokio-rs/prost/pull/785

Reply:
I missed that that was a year ago 😭 Maybe someone will want to pick up that torch though.

Reply:
> I would also tend to think runaway recursion would have to first get past the max message size limits?

I think it's pretty trivial to create a nested struct that is well below the message limit but with significant recursion.

I see what you are saying about the difficulties of configuring it at runtime if that was possible. I think I'd be ok with seeing a build time increase to the limit (say to 1000).